### DISCUSSION TITLE
[RTK] Manual Cache Updates still has gap

### DISCUSSION THREAD

**Discussion by kunyan** [reduxjs/redux-toolkit#2062] [General]
We are trying to implement a topic list system, each topic can be voted by user.

We are using RTK query to fetch all topics. and trying to use optimistic update for vote feature to improve UX .

We updated the state in `onQueryStarted`  no matter the api response is ok or not. to make sure when vote button clicked, the +1 will show up immediately. 

However, we have topic filter which could have many condition as query parameter,  all the query results had been store in different parts. 

We can not manual update all query result.......

I checked the official docs [Tradeoff](https://redux-toolkit.js.org/rtk-query/usage/cache-behavior#tradeoffs) part
Looks it is a problem which we can not resolve now.  :(

**Comment by markerikson** [comment]
I'm afraid I'm not sure what the actual question is here.  Can you clarify what you're trying to do and what isn't working?  Code would help, or even better, a CodeSandbox or [Replay recording](https://replay.io) that shows the issue.

**Reply by phryneas** [reply]
If this is just about "finding all cache entries that provide a certain thing", you can use `providesTags` together with `api.util.selectInvalidatedBy(state, tags)` to get all endpoints and associated args providing that certain tag value.

**Comment by kunyan** [comment]
Let me show more details

A topic vote system like this

| Vote      | Name | Status |
| ----------- | ----------- | ----------- |
| 102 Votes      | Feature a       | Accept |
| 10 Votes   | Feature b        | Reject |


By default ,there are no condition for the topic list. will show all topics.
The cache in state will be use `getTopics('')` with no parameter as cache key

We add a vote button in each item in the list. if user click the vote button, the votes will +1. 
| Vote      | Name | Status |
| ----------- | ----------- | ----------- |
| 103 Votes      | Feature a       | Accept |
| 10 Votes   | Feature b        | Reject |


We trying to provide a better UX. so we choosed [optimistic update](https://redux-toolkit.js.org/rtk-query/usage/manual-cache-updates#optimistic-updates).  Updated the cache data first. give a quick response to user. if the api false will revert changes.


It looks pretty good for now.
However, we have filter for the topic list. For example, we can filter topics by status.
The cache key will use `getTopics('status: accept')`.
So we have to all cache for `getTopics` no matter what parameter it is have.


`selectInvalidatedBy ` looks a good solution, we have to test if it has good performance or not. will update later
